near-sdk/src/store/vec/iter.rs

```diff
@@ -0,0 +1,158 @@
+use borsh::{BorshDeserialize, BorshSerialize};
+use core::{iter::FusedIterator, ops::Range};
+
+use super::{Vector, ERR_INDEX_OUT_OF_BOUNDS};
+use crate::env;
+
+/// An interator over references to each element in the stored vector.
+#[derive(Debug)]
+pub struct Iter<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    /// Underlying vector to iterate through
+    vec: &'a Vector<T>,
+    /// Range of indices to iterate.
+    range: Range<u32>,
+}
+
+impl<'a, T> Iter<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    pub(super) fn new(vec: &'a Vector<T>) -> Self {
+        Self { vec, range: Range { start: 0, end: vec.len() } }
+    }
+
+    /// Returns number of elements left to iterate.
+    fn remaining(&self) -> usize {
+        self.range.len()
+    }
+}
+
+impl<'a, T> Iterator for Iter<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        <Self as Iterator>::nth(self, 0)
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.remaining();
+        (remaining, Some(remaining))
+    }
+
+    fn count(self) -> usize {
+        self.remaining()
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        let idx = self.range.nth(n)?;
+        Some(self.vec.get(idx).unwrap_or_else(|| env::panic_str(ERR_INDEX_OUT_OF_BOUNDS)))
+    }
+}
+
+impl<'a, T> ExactSizeIterator for Iter<'a, T> where T: BorshSerialize + BorshDeserialize {}
+impl<'a, T> FusedIterator for Iter<'a, T> where T: BorshSerialize + BorshDeserialize {}
+
+impl<'a, T> DoubleEndedIterator for Iter<'a, T>
+where
+    T: BorshSerialize + BorshDeserialize,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        <Self as DoubleEndedIterator>::nth_back(self, 0)
+    }
+
+    fn nth_back(&mut self, n: usize) -> Option<Self::Item> {
+        let idx = self.range.nth_back(n)?;
+        Some(self.vec.get(idx).unwrap_or_else(|| env::panic_str(ERR_INDEX_OUT_OF_BOUNDS)))
+    }
+}
+
+/// An iterator over exclusive references to each element of a stored vector.
+#[derive(Debug)]
```

Comment:
Do we want the `Debug` here? Genuinely don't know but saw this in `near-sdk/src/collections/vector.rs` and thought I'd mentioned it, seeing if this is a good pattern to follow:
`#[cfg_attr(not(feature = "expensive-debug"), derive(Debug))]`
https://github.com/near/near-sdk-rs/blob/072b47779fcb219f468977bd7e727b4a83d9bd44/near-sdk/src/collections/vector.rs#L22

Reply:
Well, since derived, it would use this pattern for when the feature is enabled. There is nothing in the iterator that would be different here